core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1930,31 +1929,33 @@ private void doCheckIndex() throws IOException {
             }
         } else {
             // full checkindex
-            try (CheckIndex checkIndex = new CheckIndex(store.directory())) {
-                checkIndex.setInfoStream(out);
-                CheckIndex.Status status = checkIndex.checkIndex();
-                out.flush();
-
-                if (!status.clean) {
-                    if (state == IndexShardState.CLOSED) {
-                        // ignore if closed....
-                        return;
-                    }
-                    logger.warn("check index [failure]\n{}", os.bytes().utf8ToString());
-                    if ("fix".equals(checkIndexOnStartup)) {
-                        if (logger.isDebugEnabled()) {
-                            logger.debug("fixing index, writing new segments file ...");
+            store.runUnderMetadataLock(() -> {
```

Comment:
instead of adding this `runUnderMetadataLock` method would it make sense to just move the checkindex method into store like this `public CheckIndex.Status checkIndex()` then we can interpret the result here and run under the appropriate locks.